Frameworks/UIKit/NSLayoutManager.mm

```diff
@@ -72,105 +85,107 @@ - (void)__layoutAllText {
         while (stringIndex < lineEnd) {
             if (origin.y > containerSize.height) {
                 // Added as much text as can fit in the frame, can end here
                 return;
             }
 
-            if (!__lineHasGlyphsAfterIndex(line, stringIndex)) {
-                // Ended up without any visible glyphs to draw
-                // Caused by a line of only whitespace
+            if (stringIndex > indexOfLastVisibleChar) {
+                // Ended up without any visible glyphs to draw, caused by a line of only whitespace
                 origin = { 0.0f, origin.y + lineHeight };
-                _totalSize.height += lineHeight;
                 break;
             }
 
-            CGRect remainingRect = {};
-            CGRect rect = [container lineFragmentRectForProposedRect:CGRectMake(origin.x, origin.y, containerSize.width, lineHeight)
-                                                             atIndex:stringIndex
-                                                    writingDirection:NSWritingDirectionLeftToRight
-                                                       remainingRect:&remainingRect];
+            CGRect remainingRect{};
+            proposedRect.origin = origin;
+            const CGRect rect = [container lineFragmentRectForProposedRect:proposedRect
+                                                                   atIndex:stringIndex
+                                                          writingDirection:NSWritingDirectionLeftToRight
+                                                             remainingRect:&remainingRect];
 
             // Approximate how many characters can fit to keep from re-rendering large amounts of text
-            CFIndex lastIndex = std::min(CTLineGetStringIndexForPosition(line, { drawnWidth + rect.size.width, 0.0f }) + 1, lineEnd);
+            const CFIndex lastIndex = std::min(CTLineGetStringIndexForPosition(line, { drawnWidth + rect.size.width, 0.0f }) + 1, lineEnd);
 
             if (lastIndex == stringIndex) {
                 // Unable to fit any text in the current rect, continue to next
                 if (remainingRect.size.width > 0.0f && stringIndex < lineEnd) {
                     origin = { remainingRect.origin.x, origin.y };
                 } else {
                     origin = { 0.0f, origin.y + lineHeight };
-                    _totalSize.height += lineHeight;
                 }
 
                 continue;
             }
 
-            woc::unique_cf<CGPathRef> path{ CGPathCreateWithRect(CGRectMake(0, 0, rect.size.width, lineHeight), nullptr) };
+            frameDrawRect.size.width = rect.size.width;
+            woc::unique_cf<CGPathRef> path{ CGPathCreateWithRect(frameDrawRect, nullptr) };
             woc::unique_cf<CTFrameRef> frame{
                 CTFramesetterCreateFrame(framesetter.get(), { stringIndex, lastIndex - stringIndex }, path.get(), nullptr)
```

Comment:
I'm throwing in the towel on trying to batch these for the time being. There's a general chicken-and-egg problem re:batching here where some information required for the next line isn't known until the current one is drawn. We can estimate these like lastIndex above, but the more we estimate before grounding ourselves, the more likely it is that our estimates will diverge. Not really sure how to tackle this for the time being.